<a href="https://colab.research.google.com/github/sundeepkhatri5/Gender-Recognition-Using-Voice-Sample/blob/main/Project_ML3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true" width="500" height="240" />

# USAGE OF FACEBOOK

# Table of Contents
## Problem Statement
## Data Description
## Choosing the Tool of Analysis
## Data Collection
### 4.1 Importing the modules & setting the analysis environment
### 4.2 Loading the dataset using Pandas DataFrame
### 4.3 Pre-reading the dataset
## Pre-profiling the Dataset
## Pre-processing the Dataset
### 6.1 Assumptions
### 6.2 Check the fields with null values
### 6.3 Dropping the columns which are not required
### 6.4 Adding new fields
## Post-profiling the Dataset
## Data Normalisation
## Exploratory Data Analysis
### 9.1 Univariate
### 9.2 Bivariate
### 9.3 Multiivariate
## Model Building
### 10.1 Importing Libraries
### 10.2 Model - Verifying the assumptions for Regression
### 10.2.1 Multicollinearity
### 10.2.1 Label Distribution
### 10.2.1 Relationship between Features and Labels
### 10.3 Encoding of Categorical Features
### 10.4 Model - Linear Regression
### 10.5 Model - Decision Tree
### 10.6 Model - Decision Tree with Manual Tuning
### 10.7 Model - Decision Tree with Grid Search
### 10.8 Model - Random Forest
### 10.9 Model - Random Forest with Manual Tuning
### 10.10 Model - Random Forest with Randomized Search CV
### 10.11 Model Evaluation
## Recommendations & Actionable Insights

## Problem Statement
Facebook has become quite popular in the recent few years with people all over the world using it as a medium to share their status,thoughts and pictures with their friends. With the mobile app , it has become even more popular and easier to access. People from all age groups are connected to facebook, however there are certain differences in their way of using it such as with respect to initiating friendships and sending likes . We are trying to study the dataset provided to identify certain patterns with respect to how the users are making use of this most popular social networking app depending on their age group,gender etc.

# Desciption of the Dataset
This dataset provides a usage analysis of various facebook users such as the tenure of their account,number of friends they have,number of friendships they have initiated , the number of likes received by them through the FB website and through the mobile app .
## Column Name : Description
### userid : A numeric value uniquely identifying the user

### age : Age of the user in years

### dob_day : Day part of the user's date of birth

### dob_year : Year part of the user's date of birth

### dob_month : Month part of the user's date of birth

### gender : Gender of the user

### tenure : Number of days since the user has been on FB

### friend_count : Number of friends the user has

### friendships_initiated : Number of friendships initiated by the user

### likes : Total number of posts liked by the user

### likes_received : Total Number of likes received by user's posts

### mobile_likes : Number of posts liked by the user through mobile app

### mobile_likes_received : Number of likes received by user through mobile app

### www_likes : Number of posts liked by the user through web

### www_likes_received : Number of likes received by user through web

## CHOOSING TOOLS

In [34]:
import sklearn
import numpy as np
import pandas as pd
pd.set_option('mode.chained_assignment', None)      # To suppress pandas warnings.
pd.set_option('display.max_colwidth', -1)           # To display all the data in each column
pd.options.display.max_columns = 50                 # To display every column of the dataset in head()

import warnings
warnings.filterwarnings('ignore') 
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from collections import Counter

In [35]:
import sklearn.metrics as metrics

## IMPORTING DATASET

In [36]:
fb = pd.read_csv("https://github.com/insaid2018/Term-1/raw/master/Data/Projects/facebook_data.csv")
fb.head()

userid  age  dob_day  dob_year  dob_month  gender  tenure  friend_count  \
0  2094382  14   19       1999      11         male    266.0   0              
1  1192601  14   2        1999      11         female  6.0     0              
2  2083884  14   16       1999      11         male    13.0    0              
3  1203168  14   25       1999      12         female  93.0    0              
4  1733186  14   4        1999      12         male    82.0    0              

   friendships_initiated  likes  likes_received  mobile_likes  \
0  0                      0      0               0              
1  0                      0      0               0              
2  0                      0      0               0              
3  0                      0      0               0              
4  0                      0      0               0              

   mobile_likes_received  www_likes  www_likes_received  
0  0                      0          0                   
1  0                      0          0                   
2  0                      0          0                   
3  0                      0          0                   
4  0                      0          0

In [37]:
fb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99003 entries, 0 to 99002
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   userid                 99003 non-null  int64  
 1   age                    99003 non-null  int64  
 2   dob_day                99003 non-null  int64  
 3   dob_year               99003 non-null  int64  
 4   dob_month              99003 non-null  int64  
 5   gender                 98828 non-null  object 
 6   tenure                 99001 non-null  float64
 7   friend_count           99003 non-null  int64  
 8   friendships_initiated  99003 non-null  int64  
 9   likes                  99003 non-null  int64  
 10  likes_received         99003 non-null  int64  
 11  mobile_likes           99003 non-null  int64  
 12  mobile_likes_received  99003 non-null  int64  
 13  www_likes              99003 non-null  int64  
 14  www_likes_received     99003 non-null  int64  
dtypes:

In [38]:
fb.describe(include='all')

userid           age       dob_day      dob_year     dob_month  \
count   9.900300e+04  99003.000000  99003.000000  99003.000000  99003.000000   
unique NaN           NaN           NaN           NaN           NaN             
top    NaN           NaN           NaN           NaN           NaN             
freq   NaN           NaN           NaN           NaN           NaN             
mean    1.597045e+06  37.280224     14.530408     1975.719776   6.283365       
std     3.440592e+05  22.589748     9.015606      22.589748     3.529672       
min     1.000008e+06  13.000000     1.000000      1900.000000   1.000000       
25%     1.298806e+06  20.000000     7.000000      1963.000000   3.000000       
50%     1.596148e+06  28.000000     14.000000     1985.000000   6.000000       
75%     1.895744e+06  50.000000     22.000000     1993.000000   9.000000       
max     2.193542e+06  113.000000    31.000000     2000.000000   12.000000      

       gender        tenure  friend_count  friendships_initiated  \
count   98828  99001.000000  99003.000000  99003.000000            
unique  2     NaN           NaN           NaN                      
top     male  NaN           NaN           NaN                      
freq    58574 NaN           NaN           NaN                      
mean    NaN    537.887375    196.350787    107.452471              
std     NaN    457.649874    387.304229    188.786951              
min     NaN    0.000000      0.000000      0.000000                
25%     NaN    226.000000    31.000000     17.000000               
50%     NaN    412.000000    82.000000     46.000000               
75%     NaN    675.000000    206.000000    117.000000              
max     NaN    3139.000000   4923.000000   4144.000000             

               likes  likes_received  mobile_likes  mobile_likes_received  \
count   99003.000000  99003.000000    99003.000000  99003.000000            
unique NaN           NaN             NaN           NaN                      
top    NaN           NaN             NaN           NaN                      
freq   NaN           NaN             NaN           NaN                      
mean    156.078785    142.689363      106.116300    84.120491               
std     572.280681    1387.919613     445.252985    839.889444              
min     0.000000      0.000000        0.000000      0.000000                
25%     1.000000      1.000000        0.000000      0.000000                
50%     11.000000     8.000000        4.000000      4.000000                
75%     81.000000     59.000000       46.000000     33.000000               
max     25111.000000  261197.000000   25111.000000  138561.000000           

           www_likes  www_likes_received  
count   99003.000000  99003.000000        
unique NaN           NaN                  
top    NaN           NaN                  
freq   NaN           NaN                  
mean    49.962425     58.568831           
std     285.560152    601.416348          
min     0.000000      0.000000            
25%     0.000000      0.000000            
50%     0.000000      2.000000            
75%     7.000000      20.000000           
max     14865.000000  129953.000000

In [39]:
fb['gender'] = fb['gender'].fillna(fb['gender'].mode()[0])


In [40]:
fb['tenure']=fb['tenure'].fillna(fb['tenure'].mode()[0])


In [41]:
fb.describe()

userid           age       dob_day      dob_year     dob_month  \
count  9.900300e+04  99003.000000  99003.000000  99003.000000  99003.000000   
mean   1.597045e+06  37.280224     14.530408     1975.719776   6.283365       
std    3.440592e+05  22.589748     9.015606      22.589748     3.529672       
min    1.000008e+06  13.000000     1.000000      1900.000000   1.000000       
25%    1.298806e+06  20.000000     7.000000      1963.000000   3.000000       
50%    1.596148e+06  28.000000     14.000000     1985.000000   6.000000       
75%    1.895744e+06  50.000000     22.000000     1993.000000   9.000000       
max    2.193542e+06  113.000000    31.000000     2000.000000   12.000000      

             tenure  friend_count  friendships_initiated         likes  \
count  99003.000000  99003.000000  99003.000000           99003.000000   
mean   537.882569    196.350787    107.452471             156.078785     
std    457.646500    387.304229    188.786951             572.280681     
min    0.000000      0.000000      0.000000               0.000000       
25%    226.000000    31.000000     17.000000              1.000000       
50%    412.000000    82.000000     46.000000              11.000000      
75%    675.000000    206.000000    117.000000             81.000000      
max    3139.000000   4923.000000   4144.000000            25111.000000   

       likes_received  mobile_likes  mobile_likes_received     www_likes  \
count  99003.000000    99003.000000  99003.000000           99003.000000   
mean   142.689363      106.116300    84.120491              49.962425      
std    1387.919613     445.252985    839.889444             285.560152     
min    0.000000        0.000000      0.000000               0.000000       
25%    1.000000        0.000000      0.000000               0.000000       
50%    8.000000        4.000000      4.000000               0.000000       
75%    59.000000       46.000000     33.000000              7.000000       
max    261197.000000   25111.000000  138561.000000          14865.000000   

       www_likes_received  
count  99003.000000        
mean   58.568831           
std    601.416348          
min    0.000000            
25%    0.000000            
50%    2.000000            
75%    20.000000           
max    129953.000000

In [42]:

fb=fb.drop('userid',axis=1, inplace=False)

In [43]:
fb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99003 entries, 0 to 99002
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   age                    99003 non-null  int64  
 1   dob_day                99003 non-null  int64  
 2   dob_year               99003 non-null  int64  
 3   dob_month              99003 non-null  int64  
 4   gender                 99003 non-null  object 
 5   tenure                 99003 non-null  float64
 6   friend_count           99003 non-null  int64  
 7   friendships_initiated  99003 non-null  int64  
 8   likes                  99003 non-null  int64  
 9   likes_received         99003 non-null  int64  
 10  mobile_likes           99003 non-null  int64  
 11  mobile_likes_received  99003 non-null  int64  
 12  www_likes              99003 non-null  int64  
 13  www_likes_received     99003 non-null  int64  
dtypes: float64(1), int64(12), object(1)
memory usage: 10.6

In [44]:

# Extract target variable
targets = pd.DataFrame(fb['gender'])
# Extract independent variables
features = fb.drop('gender',axis = 1)

from sklearn.model_selection import train_test_split

# Split into 80% training and 20% testing set
X_train, X_test, y_train, y_test = train_test_split(features, targets, test_size = 0.2, random_state = 1)

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(79202, 13)
(19801, 13)
(79202, 1)
(19801, 1)


In [45]:

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
# While scaling the test data, mean and standard deviation of train data
# should be considered.
X_test = sc.transform(X_test)

In [46]:
# Encoding the Dependent Variable
from sklearn.preprocessing import LabelEncoder
labelencoder_y_train = LabelEncoder()
y_train = labelencoder_y_train.fit_transform(y_train)
labelencoder_y_test = LabelEncoder()
y_test = labelencoder_y_test.fit_transform(y_test)

In [47]:
y_train[:10]

array([0, 0, 0, 0, 1, 0, 1, 1, 0, 0])

In [ ]:
# # Logistic Regression
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()

In [ ]:
fb = fb.reset_index()

In [ ]:
Model1 = logreg.fit(X_train,y_train)

In [ ]:

y_train_pred_lm = Model1.predict(X_train)

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,recall_score

In [ ]:
accuracy_score(y_train,y_train_pred_lm)

0.6409055326885684

In [ ]:
cm_train_lm = confusion_matrix(y_train,y_train_pred_lm)
print(cm_train_lm)

[[ 6735 25534]
 [ 2907 44026]]


In [ ]:
y_test_pred_lm = Model1.predict(X_test)

In [ ]:
accuracy_score(y_test,y_test_pred_lm)

0.639917175900207

In [ ]:

recall_score(y_test,y_test_pred_lm)

0.936357481381178

In [ ]:
cm_test_lm = confusion_matrix(y_test,y_test_pred_lm)
print(cm_test_lm)

[[ 1607  6378]
 [  752 11064]]


# In case if i predicted Male and it turned out to be female, it would mean a loss of business. To minimise False Negative we need to maximise recall

In [ ]:
recall_score(y_test,y_test_pred_lm)

0.936357481381178

# So the algorithm has a high degree of recall ie of all gender predictions only around 6.35% are wrong hence less loss of business by wrong promotional campaigns
 

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR


#Now let's try pycaret and compare all classification models and see what gives me the best solution

In [3]:
!pip install pycaret

     |████████████████████████████████| 256kB 5.3MB/s 
     |████████████████████████████████| 6.8MB 7.4MB/s 
     |████████████████████████████████| 174kB 52.3MB/s 
     |████████████████████████████████| 65.8MB 53kB/s 
     |████████████████████████████████| 245kB 53.2MB/s 
     |████████████████████████████████| 276kB 52.9MB/s 
     |████████████████████████████████| 102kB 12.6MB/s 
     |████████████████████████████████| 157.5MB 80kB/s 
     |████████████████████████████████| 14.2MB 49.1MB/s 
     |████████████████████████████████| 1.8MB 54.3MB/s 
     |████████████████████████████████| 1.6MB 38.6MB/s 
     |████████████████████████████████| 61kB 7.8MB/s 
     |████████████████████████████████| 604kB 39.8MB/s 
     |████████████████████████████████| 81kB 227kB/s 
     |████████████████████████████████| 3.1MB 36.1MB/s 
     |████████████████████████████████| 81kB 10.8MB/s 
     |████████████████████████████████| 2.1MB 45.7MB/s 
     |████████████████████████████████| 163kB 58.4MB/s 

In [4]:
import pycaret
import numpy as np
import pandas as pd

In [6]:
fb = pd.read_csv("https://github.com/insaid2018/Term-1/raw/master/Data/Projects/facebook_data.csv")
fb.head()

userid  age  dob_day  dob_year  dob_month  gender  tenure  friend_count  \
0  2094382   14       19      1999         11    male   266.0             0   
1  1192601   14        2      1999         11  female     6.0             0   
2  2083884   14       16      1999         11    male    13.0             0   
3  1203168   14       25      1999         12  female    93.0             0   
4  1733186   14        4      1999         12    male    82.0             0   

   friendships_initiated  likes  likes_received  mobile_likes  \
0                      0      0               0             0   
1                      0      0               0             0   
2                      0      0               0             0   
3                      0      0               0             0   
4                      0      0               0             0   

   mobile_likes_received  www_likes  www_likes_received  
0                      0          0                   0  
1                      0          0                   0  
2                      0          0                   0  
3                      0          0                   0  
4                      0          0                   0

In [7]:
# Extract target variable
targets = pd.DataFrame(fb['gender'])
# Extract independent variables
features = fb.drop('gender',axis = 1)

In [49]:
from pycaret.classification import *

In [51]:
clf1 = setup(data = fb, 
             target = 'gender',
             numeric_imputation = 'mean',
             categorical_features = ['dob_month'])

Description               Value
0   session_id                              1454              
1   Target                                  gender            
2   Target Type                             Binary            
3   Label Encoded                           female: 0, male: 1
4   Original Data                           (99003, 14)       
5   Missing Values                          False             
6   Numeric Features                        12                
7   Categorical Features                    1                 
8   Ordinal Features                        False             
9   High Cardinality Features               False             
10  High Cardinality Method                 None              
11  Transformed Train Set                   (69302, 23)       
12  Transformed Test Set                    (29701, 23)       
13  Shuffle Train-Test                      True              
14  Stratify Train-Test                     False             
15  Fold Generator                          StratifiedKFold   
16  Fold Number                             10                
17  CPU Jobs                                -1                
18  Use GPU                                 False             
19  Log Experiment                          False             
20  Experiment Name                         clf-default-name  
21  USI                                     5bd1              
22  Imputation Type                         simple            
23  Iterative Imputation Iteration          None              
24  Numeric Imputer                         mean              
25  Iterative Imputation Numeric Model      None              
26  Categorical Imputer                     constant          
27  Iterative Imputation Categorical Model  None              
28  Unknown Categoricals Handling           least_frequent    
29  Normalize                               False             
30  Normalize Method                        None              
31  Transformation                          False             
32  Transformation Method                   None              
33  PCA                                     False             
34  PCA Method                              None              
35  PCA Components                          None              
36  Ignore Low Variance                     False             
37  Combine Rare Levels                     False             
38  Rare Level Threshold                    None              
39  Numeric Binning                         False             
40  Remove Outliers                         False             
41  Outliers Threshold                      None              
42  Remove Multicollinearity                False             
43  Multicollinearity Threshold             None              
44  Clustering                              False             
45  Clustering Iteration                    None              
46  Polynomial Features                     False             
47  Polynomial Degree                       None              
48  Trignometry Features                    False             
49  Polynomial Threshold                    None              
50  Group Features                          False             
51  Feature Selection                       False             
52  Features Selection Threshold            None              
53  Feature Interaction                     False             
54  Feature Ratio                           False             
55  Interaction Threshold                   None              
56  Fix Imbalance                           False             
57  Fix Imbalance Method                    SMOTE

In [52]:
compare_models()

Model  Accuracy     AUC  Recall   Prec.  \
catboost  CatBoost Classifier              0.6996    0.7459  0.8266  0.7126   
lightgbm  Light Gradient Boosting Machine  0.6977    0.7432  0.8284  0.7101   
gbc       Gradient Boosting Classifier     0.6951    0.7356  0.8346  0.7052   
xgboost   Extreme Gradient Boosting        0.6939    0.7361  0.8176  0.7100   
rf        Random Forest Classifier         0.6883    0.7233  0.8089  0.7073   
ada       Ada Boost Classifier             0.6871    0.7223  0.8337  0.6975   
et        Extra Trees Classifier           0.6820    0.7112  0.8054  0.7021   
knn       K Neighbors Classifier           0.6444    0.6617  0.7491  0.6822   
lr        Logistic Regression              0.6434    0.6709  0.9468  0.6333   
lda       Linear Discriminant Analysis     0.6245    0.6504  0.9422  0.6208   
ridge     Ridge Classifier                 0.6240    0.0000  0.9442  0.6201   
nb        Naive Bayes                      0.6170    0.6228  0.9722  0.6113   
dt        Decision Tree Classifier         0.5997    0.5865  0.6574  0.6640   
qda       Quadratic Discriminant Analysis  0.4966    0.5168  0.4488  0.4805   
svm       SVM - Linear Kernel              0.4725    0.0000  0.2420  0.6400   

              F1   Kappa     MCC  TT (Sec)  
catboost  0.7654  0.3542  0.3618  27.696    
lightgbm  0.7647  0.3491  0.3573  0.794     
gbc       0.7644  0.3408  0.3507  10.118    
xgboost   0.7600  0.3432  0.3499  12.617    
rf        0.7547  0.3325  0.3382  10.420    
ada       0.7595  0.3215  0.3322  2.453     
et        0.7502  0.3185  0.3243  9.208     
knn       0.7141  0.2465  0.2484  1.766     
lr        0.7589  0.1673  0.2296  0.758     
lda       0.7484  0.1181  0.1703  0.269     
ridge     0.7486  0.1158  0.1690  0.088     
nb        0.7506  0.0833  0.1523  0.073     
dt        0.6607  0.1727  0.1727  0.666     
qda       0.4111  0.0155  0.0188  0.122     
svm       0.2580  0.0506  0.0727  0.464

#Therefore as we see Naive Bayes offers the highest recall value amongst all values, even higher than logistic regression

## Create Model

In [54]:
nb = create_model('nb')      

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
0     0.6204    0.6273  0.9715  0.6136  0.7521  0.0930  0.1652
1     0.6218    0.6402  0.9732  0.6143  0.7532  0.0963  0.1716
2     0.6097    0.6059  0.9693  0.6070  0.7465  0.0638  0.1208
3     0.6150    0.6265  0.9713  0.6102  0.7495  0.0779  0.1437
4     0.6214    0.6269  0.9735  0.6140  0.7530  0.0946  0.1697
5     0.6188    0.6225  0.9725  0.6124  0.7515  0.0881  0.1597
6     0.6137    0.6159  0.9720  0.6092  0.7489  0.0742  0.1393
7     0.6130    0.6222  0.9725  0.6086  0.7487  0.0719  0.1365
8     0.6162    0.6220  0.9696  0.6111  0.7497  0.0823  0.1481
9     0.6205    0.6188  0.9764  0.6129  0.7531  0.0911  0.1688
Mean  0.6170    0.6228  0.9722  0.6113  0.7506  0.0833  0.1523
SD    0.0039    0.0084  0.0019  0.0024  0.0022  0.0105  0.0163

In [55]:
tuned_nb = tune_model(nb)

Accuracy     AUC  Recall   Prec.      F1   Kappa     MCC
0     0.6204    0.6281  0.9715  0.6136  0.7521  0.0930  0.1652
1     0.6217    0.6409  0.9732  0.6142  0.7531  0.0959  0.1711
2     0.6098    0.6068  0.9693  0.6071  0.7466  0.0642  0.1214
3     0.6149    0.6272  0.9713  0.6101  0.7494  0.0775  0.1432
4     0.6214    0.6272  0.9735  0.6140  0.7530  0.0946  0.1697
5     0.6186    0.6231  0.9725  0.6123  0.7514  0.0877  0.1591
6     0.6137    0.6166  0.9720  0.6092  0.7489  0.0742  0.1393
7     0.6127    0.6231  0.9725  0.6084  0.7485  0.0711  0.1353
8     0.6160    0.6228  0.9696  0.6110  0.7496  0.0819  0.1475
9     0.6205    0.6195  0.9764  0.6129  0.7531  0.0911  0.1688
Mean  0.6170    0.6235  0.9722  0.6113  0.7506  0.0831  0.1521
SD    0.0039    0.0083  0.0019  0.0024  0.0022  0.0104  0.0163

In [56]:
evaluate_model(tuned_nb)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

# Thus the tuned Naive Bayes model offers a recall value of a mean of 97.22% and a std deviation of 0.19%